### PR TITLE
Fix bugs reported in #1338

### PR DESCRIPTION
**Pull request by leobalter**
cc @anba

### CODE REVIEW

**Comment by anba** [comment]
Looks good to me, down to only five failures:

- test/language/statements/class/fields-computed-name-static-propname-prototype.js
- test/language/expressions/class/fields-computed-name-static-propname-prototype.js

These two tests should expect a TypeError, because the tests attempt to redefine the non-configurable, non-writable `prototoype` property on a class. So basically it's similar to `class C { static ["prototype"](){} }`.

- test/language/statements/class/privatefieldget-success-5.js

The expected value should be the number `42`, not the string `'42'` in 88a2e077297d96d77ca188236ed0e001b68b4734.

- test/language/statements/class/privatefieldget-typeerror-3.js
- test/language/statements/class/privatefieldset-typeerror-3.js

&rarr; https://github.com/tc39/test262/issues/1338#issuecomment-342639706

**Comment by leobalter** [comment]
> These two tests should expect a TypeError, because the tests attempt to redefine the non-configurable, non-writable prototoype property on a class. So basically it's similar to class C { static ["prototype"](){} }.

oops.

I need to take this off these cases off the generation tool. Not an issue. Thanks for catching it.

File: test/language/statements/class/privatefieldset-typeerror-3.js [hunk 1]
```diff
@@ -40,16 +41,22 @@ class Outer {
...
 
-assert.throws(TypeError, function() {
-  // when f() is called, the private field will not be found in Inner's `this`
-  i.f();
-})
+assert.sameValue(outer.value(), 42);
+
+i.f();
+
+assert.sameValue(outer.value(), 1, "value is set from inner class instance");
```
Comment (leobalter) [inline test/language/statements/class/privatefieldset-typeerror-3.js]: this feels so weird. I wonder if it is wrong or if it should be wrong.

**Comment by anba** [comment]
d24e325e1a9657a9591cb8aa80276f0929307236 still needs some tweaking:
- The TypeError is already thrown during the ClassDeclaration/Expression evaluation, not when an instance of the class is instantiated.

7f94fffbafe54e9ac3f127a7b1b7fab227a1d286
- There's still something wrong here. The assertions should probably be:
```js
var outer = new Outer();
var Inner = outer.innerclass();
var i = new Inner();

assert.sameValue(outer.value(), 42);
assert.sameValue(i.value(), 42);

i.f();

assert.sameValue(outer.value(), 42, "value is set from inner class instance");
assert.sameValue(i.value(), 1, "value is set from inner class instance");
```

**Comment by leobalter** [comment]
> The TypeError is already thrown during the ClassDeclaration/Expression evaluation, not when an instance of the class is instantiated.

https://tc39.github.io/ecma262/#sec-object-initializer-runtime-semantics-evaluation

Nice! This is the evaluation I was looking for.

---

```js
assert.sameValue(outer.value(), 42, "value is set from inner class instance");
assert.sameValue(i.value(), 1, "value is set from inner class instance");
```

I applied this fix, but I still think the inner class should only refer to `#x` through `super.#x`, not `this.#x`.

**Comment by leobalter** [comment]
Thanks for the review, @anba! I hope we can get rid of these bugs for good. :)

**Comment by leobalter** [comment]
Assuming these tests are certainly more correct than the ones with bugs, I'll land this and make myself available to fix anything else further.

**Pull request merged**